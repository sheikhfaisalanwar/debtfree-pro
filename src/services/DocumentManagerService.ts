import { DocumentUploadService, UploadedDocument } from './DocumentUploadService';
import { DocumentValidationService, ValidationResult } from './DocumentValidationService';
import { Statement, StatementTransaction, StatementPayment } from '../types/Statement';
import { DataStoreService } from './DataStoreService';

export interface ProcessedDocument extends UploadedDocument {
  validationResult?: ValidationResult;
  extractedData?: Statement;
  processingError?: string;
}

export interface DocumentProcessingResult {
  success: boolean;
  document?: ProcessedDocument;
  statement?: Statement;
  error?: string;
}

export class DocumentManagerService {
  static async uploadAndProcessDocument(debtId?: string): Promise<DocumentProcessingResult> {
    try {
      const uploadResult = await DocumentUploadService.pickDocument();
      
      if (!uploadResult.success || !uploadResult.document) {
        return {
          success: false,
          error: uploadResult.error
        };
      }

      const document = uploadResult.document;
      if (debtId) {
        document.debtId = debtId;
      }

      const processingResult = await this.processDocument(document);
      
      return {
        success: processingResult.success,
        document: processingResult.document,
        statement: processingResult.statement,
        error: processingResult.error
      };
    } catch (error) {
      return {
        success: false,
        error: `Upload and processing failed: ${error}`
      };
    }
  }

  static async processDocument(document: UploadedDocument): Promise<DocumentProcessingResult> {
    try {
      let content = '';
      
      if (document.fileType === 'csv') {
        content = await DocumentUploadService.readDocumentContent(document);
      }

      const validationResult = await DocumentValidationService.validateDocument(document, content);
      
      const processedDoc: ProcessedDocument = {
        ...document,
        validationResult
      };

      if (!validationResult.isValid) {
        processedDoc.processingError = 'Document validation failed';
        return {
          success: false,
          document: processedDoc,
          error: `Validation failed: ${validationResult.errors.join(', ')}`
        };
      }

      if (document.fileType === 'csv' && content) {
        const statement = await this.extractStatementFromCSV(document, content, validationResult);
        processedDoc.extractedData = statement;
        processedDoc.processed = true;

        await DataStoreService.addStatement(statement);

        return {
          success: true,
          document: processedDoc,
          statement
        };
      }

      processedDoc.processed = true;
      
      return {
        success: true,
        document: processedDoc
      };
    } catch (error) {
      return {
        success: false,
        document: {
          ...document,
          processingError: `Processing failed: ${error}`
        },
        error: `Document processing failed: ${error}`
      };
    }
  }

  static async getProcessedDocuments(debtId?: string): Promise<ProcessedDocument[]> {
    try {
      const allDocuments = await DocumentUploadService.getUploadedDocuments();
      
      return allDocuments
        .filter(doc => !debtId || doc.debtId === debtId)
        .map(doc => doc as ProcessedDocument);
    } catch (error) {
      console.error('Failed to get processed documents:', error);
      return [];
    }
  }

  static async deleteDocument(documentId: string): Promise<boolean> {
    return await DocumentUploadService.deleteDocument(documentId);
  }

  private static async extractStatementFromCSV(
    document: UploadedDocument, 
    content: string, 
    _validationResult: ValidationResult
  ): Promise<Statement> {
    const lines = content.split('\n').filter(line => line.trim());
    const headers = this.parseCSVRow(lines[0]).map(h => h.toLowerCase().trim());
    
    const transactions: StatementTransaction[] = [];
    const payments: StatementPayment[] = [];
    let balance = 0;
    let minimumPayment = 0;
    let interestCharged = 0;

    for (let i = 1; i < lines.length; i++) {
      const values = this.parseCSVRow(lines[i]);
      
      if (values.length === headers.length) {
        const row: {[key: string]: string} = {};
        headers.forEach((header, index) => {
          row[header] = values[index].trim();
        });

        const transaction = this.parseTransactionRow(row, headers);
        if (transaction) {
          if (transaction.amount < 0) {
            payments.push({
              date: transaction.date,
              amount: Math.abs(transaction.amount),
              description: transaction.description
            });
          } else {
            transactions.push(transaction);
          }
        }
      }
    }

    if (transactions.length > 0) {
      balance = transactions.reduce((sum, t) => sum + t.amount, 0);
    }

    const statement: Statement = {
      id: `stmt_${document.id}`,
      debtId: document.debtId || 'unknown',
      statementDate: new Date(),
      balance,
      minimumPayment,
      dueDate: new Date(Date.now() + 30 * 24 * 60 * 60 * 1000), // 30 days from now
      interestCharged,
      payments,
      purchases: transactions,
      fileName: document.fileName,
      imported: new Date()
    };

    return statement;
  }

  private static parseTransactionRow(row: {[key: string]: string}, headers: string[]): StatementTransaction | null {
    try {
      const dateField = this.findFieldByPatterns(row, ['date', 'transaction_date', 'posted_date']);
      const amountField = this.findFieldByPatterns(row, ['amount', 'transaction_amount', 'debit', 'credit']);
      const descriptionField = this.findFieldByPatterns(row, ['description', 'merchant', 'payee']);

      if (!dateField || !amountField || !descriptionField) {
        return null;
      }

      const date = new Date(row[dateField]);
      if (isNaN(date.getTime())) {
        return null;
      }

      let amount = parseFloat(row[amountField].replace(/[$,]/g, ''));
      
      if (headers.includes('debit') && headers.includes('credit')) {
        const debit = parseFloat((row.debit || '0').replace(/[$,]/g, ''));
        const credit = parseFloat((row.credit || '0').replace(/[$,]/g, ''));
        amount = credit > 0 ? -credit : debit;
      }

      if (isNaN(amount)) {
        return null;
      }

      return {
        date,
        amount,
        description: row[descriptionField],
        category: this.categorizeTransaction(row[descriptionField])
      };
    } catch (error) {
      return null;
    }
  }

  private static findFieldByPatterns(row: {[key: string]: string}, patterns: string[]): string | null {
    const keys = Object.keys(row);
    
    for (const pattern of patterns) {
      const found = keys.find(key => key.toLowerCase().includes(pattern));
      if (found) return found;
    }
    
    return null;
  }

  private static categorizeTransaction(description: string): string {
    const desc = description.toLowerCase();
    
    if (desc.includes('grocery') || desc.includes('food') || desc.includes('restaurant')) {
      return 'Food & Dining';
    }
    if (desc.includes('gas') || desc.includes('fuel') || desc.includes('auto')) {
      return 'Transportation';
    }
    if (desc.includes('store') || desc.includes('retail') || desc.includes('shop')) {
      return 'Shopping';
    }
    if (desc.includes('payment') || desc.includes('transfer')) {
      return 'Payment';
    }
    
    return 'Other';
  }

  private static parseCSVRow(row: string): string[] {
    const result: string[] = [];
    let current = '';
    let inQuotes = false;

    for (let i = 0; i < row.length; i++) {
      const char = row[i];
      const nextChar = row[i + 1];

      if (char === '"' && !inQuotes) {
        inQuotes = true;
      } else if (char === '"' && inQuotes && nextChar === '"') {
        current += '"';
        i++;
      } else if (char === '"' && inQuotes) {
        inQuotes = false;
      } else if (char === ',' && !inQuotes) {
        result.push(current);
        current = '';
      } else {
        current += char;
      }
    }

    result.push(current);
    return result;
  }

  static getDocumentSummary(document: ProcessedDocument): string {
    const parts: string[] = [];
    
    parts.push(`📄 ${document.fileName}`);
    parts.push(`📊 ${document.fileType.toUpperCase()}`);
    parts.push(`📅 ${document.uploadDate.toLocaleDateString()}`);
    
    if (document.validationResult) {
      if (document.validationResult.isValid) {
        parts.push('✅ Valid');
      } else {
        parts.push('❌ Invalid');
      }
      
      if (document.validationResult.detectedType) {
        parts.push(`🏷️ ${document.validationResult.detectedType.replace('_', ' ')}`);
      }
    }
    
    if (document.extractedData) {
      const txCount = document.extractedData.purchases.length + document.extractedData.payments.length;
      parts.push(`💳 ${txCount} transactions`);
    }
    
    return parts.join(' • ');
  }
}