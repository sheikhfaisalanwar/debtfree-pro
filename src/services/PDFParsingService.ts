// No filesystem dependencies - PDFs require manual entry in React Native

export interface PDFTextData {
  text: string;
  pages: number;
  info: any;
}

export interface CreditCardStatementData {
  previousBalance?: number;
  interest?: number;
  payments?: number;
  purchases?: number;
  interestRate?: number;
  statementDate?: Date;
  dueDate?: Date;
  minimumPayment?: number;
  creditLimit?: number;
  availableCredit?: number;
}

export class PDFParsingService {
  /**
   * Extract text content from PDF file
   * NOTE: PDF text extraction is not available in React Native with open-source libraries.
   * This method provides a fallback that indicates manual entry is required.
   */
  static async extractTextFromPDF(filePath: string): Promise<PDFTextData> {
    try {
      // Handle null/undefined filePath safely
      let filename = 'statement.pdf';
      
      if (filePath && typeof filePath === 'string' && filePath.length > 0) {
        // Extract filename from path for user-friendly display
        // Handle both forward slashes (Unix/iOS) and backslashes (Windows)
        const pathSeparators = /[/\\]/;
        const pathParts = filePath.split(pathSeparators);
        filename = pathParts.pop() || 'statement.pdf';
      }
      
      // Since we can't extract text from PDFs in React Native with open-source libraries,
      // we return a placeholder that indicates manual data entry is needed
      const placeholderText = `PDF_REQUIRES_MANUAL_ENTRY:${filename}`;
      
      return {
        text: placeholderText,
        pages: 1, // Unknown page count
        info: {
          filename,
          requiresManualEntry: true
        }
      };
    } catch (error) {
      console.error('Failed to process PDF:', error);
      throw new Error(`PDF processing failed: ${error instanceof Error ? error.message : 'Unknown error'}`);
    }
  }

  /**
   * Parse credit card statement data from PDF text
   */
  static parseCreditCardStatement(pdfText: string): CreditCardStatementData {
    const statementData: CreditCardStatementData = {};
    
    // Check if this is a manual entry placeholder
    if (pdfText.startsWith('PDF_REQUIRES_MANUAL_ENTRY:')) {
      // Return empty data object - user will need to enter data manually
      return statementData;
    }
    
    // Normalize text for easier parsing
    const normalizedText = pdfText.replace(/\s+/g, ' ').toUpperCase();
    
    // Extract Previous Balance
    statementData.previousBalance = this.extractAmount(normalizedText, [
      /PREVIOUS\s+(?:STATEMENT\s+)?BALANCE[:\s]+\$?([\d,]+\.?\d*)/,
      /BEGINNING\s+BALANCE[:\s]+\$?([\d,]+\.?\d*)/,
      /OPENING\s+BALANCE[:\s]+\$?([\d,]+\.?\d*)/
    ]);

    // Extract Interest/Finance Charges
    statementData.interest = this.extractAmount(normalizedText, [
      /INTEREST\s+CHARGED?[:\s]+\$?([\d,]+\.?\d*)/,
      /FINANCE\s+CHARGE[:\s]+\$?([\d,]+\.?\d*)/,
      /INTEREST\s+FEES?[:\s]+\$?([\d,]+\.?\d*)/
    ]);

    // Extract Payments
    statementData.payments = this.extractAmount(normalizedText, [
      /PAYMENTS?(?:\s+AND\s+OTHER\s+CREDITS?)?[:\s]+\$?([\d,]+\.?\d*)/,
      /PAYMENT\s+RECEIVED[:\s]+\$?([\d,]+\.?\d*)/,
      /CREDITS?[:\s]+\$?([\d,]+\.?\d*)/
    ]);

    // Extract Purchases
    statementData.purchases = this.extractAmount(normalizedText, [
      /PURCHASES?(?:\s+AND\s+ADJUSTMENTS?)?[:\s]+\$?([\d,]+\.?\d*)/,
      /NEW\s+PURCHASES?[:\s]+\$?([\d,]+\.?\d*)/,
      /TRANSACTIONS?[:\s]+\$?([\d,]+\.?\d*)/
    ]);

    // Extract Interest Rate
    statementData.interestRate = this.extractRate(normalizedText, [
      /ANNUAL\s+PERCENTAGE\s+RATE\s+FOR\s+PURCHASES[:\s]+([\d.]+)%/,
      /PURCHASE\s+APR[:\s]+([\d.]+)%/,
      /ANNUAL\s+PERCENTAGE\s+RATE[:\s]+([\d.]+)%/,
      /INTEREST\s+RATE[:\s]+([\d.]+)%/,
      /APR[:\s]+([\d.]+)%/
    ]);

    // Extract Minimum Payment
    statementData.minimumPayment = this.extractAmount(normalizedText, [
      /MINIMUM\s+PAYMENT\s+DUE[:\s]+\$?([\d,]+\.?\d*)/,
      /MINIMUM\s+PAYMENT[:\s]+\$?([\d,]+\.?\d*)/,
      /MINIMUM\s+DUE[:\s]+\$?([\d,]+\.?\d*)/,
      /PAYMENT\s+DUE[:\s]+\$?([\d,]+\.?\d*)/
    ]);

    // Extract Credit Limit
    statementData.creditLimit = this.extractAmount(normalizedText, [
      /CREDIT\s+LINE[:\s]+\$?([\d,]+\.?\d*)/,
      /CREDIT\s+LIMIT[:\s]+\$?([\d,]+\.?\d*)/,
      /TOTAL\s+CREDIT\s+LINE[:\s]+\$?([\d,]+\.?\d*)/
    ]);

    // Extract Available Credit
    statementData.availableCredit = this.extractAmount(normalizedText, [
      /AVAILABLE\s+CREDIT[:\s]+\$?([\d,]+\.?\d*)/,
      /CREDIT\s+AVAILABLE[:\s]+\$?([\d,]+\.?\d*)/
    ]);

    // Extract Statement Date
    statementData.statementDate = this.extractDate(normalizedText, [
      /STATEMENT\s+DATE[:\s]+(\d{1,2}[\/\-]\d{1,2}[\/\-]\d{2,4})/,
      /CLOSING\s+DATE[:\s]+(\d{1,2}[\/\-]\d{1,2}[\/\-]\d{2,4})/,
      /CYCLE\s+ENDING[:\s]+(\d{1,2}[\/\-]\d{1,2}[\/\-]\d{2,4})/
    ]);

    // Extract Due Date
    statementData.dueDate = this.extractDate(normalizedText, [
      /DUE\s+DATE[:\s]+(\d{1,2}[\/\-]\d{1,2}[\/\-]\d{2,4})/,
      /PAYMENT\s+DUE\s+DATE[:\s]+(\d{1,2}[\/\-]\d{1,2}[\/\-]\d{2,4})/,
      /PAYMENT\s+DUE[:\s]+(\d{1,2}[\/\-]\d{1,2}[\/\-]\d{2,4})/
    ]);

    return statementData;
  }

  /**
   * Extract amount using regex patterns
   */
  private static extractAmount(text: string, patterns: RegExp[]): number | undefined {
    for (const pattern of patterns) {
      const match = text.match(pattern);
      if (match && match[1]) {
        const cleanAmount = match[1].replace(/,/g, '');
        const amount = parseFloat(cleanAmount);
        if (!isNaN(amount)) {
          return amount;
        }
      }
    }
    return undefined;
  }

  /**
   * Extract percentage rate using regex patterns
   */
  private static extractRate(text: string, patterns: RegExp[]): number | undefined {
    for (const pattern of patterns) {
      const match = text.match(pattern);
      if (match && match[1]) {
        const rate = parseFloat(match[1]);
        if (!isNaN(rate)) {
          return rate;
        }
      }
    }
    return undefined;
  }

  /**
   * Extract date using regex patterns with improved parsing
   */
  private static extractDate(text: string, patterns: RegExp[]): Date | undefined {
    for (const pattern of patterns) {
      const match = text.match(pattern);
      if (match && match[1]) {
        try {
          // Normalize date string to handle different formats
          let dateString = match[1].trim();
          
          // Convert MM-DD-YYYY to MM/DD/YYYY
          dateString = dateString.replace(/-/g, '/');
          
          // Parse the date
          const date = new Date(dateString);
          
          // Validate the date is actually valid and reasonable
          if (!isNaN(date.getTime()) && this.isReasonableDate(date)) {
            return date;
          }
        } catch {
          // Continue to next pattern
        }
      }
    }
    return undefined;
  }

  /**
   * Validate that a date is reasonable for a financial statement
   */
  private static isReasonableDate(date: Date): boolean {
    const now = new Date();
    const fiveYearsAgo = new Date(now.getFullYear() - 5, now.getMonth(), now.getDate());
    const oneYearFromNow = new Date(now.getFullYear() + 1, now.getMonth(), now.getDate());
    
    // Date should be within a reasonable range for financial statements
    return date >= fiveYearsAgo && date <= oneYearFromNow;
  }

  /**
   * Validate extracted credit card statement data
   */
  static validateStatementData(data: CreditCardStatementData): { isValid: boolean; errors: string[] } {
    const errors: string[] = [];

    // Check if we extracted any meaningful data
    const hasData = Object.values(data).some(value => value !== undefined);
    if (!hasData) {
      errors.push('No recognizable credit card statement data found in PDF');
    }

    // Validate amounts are positive
    const amounts = ['previousBalance', 'interest', 'payments', 'purchases', 'minimumPayment', 'creditLimit', 'availableCredit'] as const;
    for (const field of amounts) {
      const value = data[field];
      if (value !== undefined && (value < 0 || isNaN(value))) {
        errors.push(`Invalid ${field}: ${value}`);
      }
    }

    // Validate interest rate is reasonable
    if (data.interestRate !== undefined && (data.interestRate < 0 || data.interestRate > 50)) {
      errors.push(`Interest rate seems unrealistic: ${data.interestRate}%`);
    }

    // Validate credit limit vs available credit
    if (data.creditLimit !== undefined && data.availableCredit !== undefined) {
      if (data.availableCredit > data.creditLimit) {
        errors.push('Available credit cannot exceed credit limit');
      }
    }

    return {
      isValid: errors.length === 0,
      errors
    };
  }

  /**
   * Get a summary of extracted data for user review
   */
  static getDataSummary(data: CreditCardStatementData): string {
    const summary: string[] = [];
    
    if (data.previousBalance !== undefined) {
      summary.push(`Previous Balance: $${data.previousBalance.toFixed(2)}`);
    }
    if (data.purchases !== undefined) {
      summary.push(`Purchases: $${data.purchases.toFixed(2)}`);
    }
    if (data.payments !== undefined) {
      summary.push(`Payments: $${data.payments.toFixed(2)}`);
    }
    if (data.interest !== undefined) {
      summary.push(`Interest/Fees: $${data.interest.toFixed(2)}`);
    }
    if (data.interestRate !== undefined) {
      summary.push(`Interest Rate: ${data.interestRate}%`);
    }
    if (data.creditLimit !== undefined) {
      summary.push(`Credit Limit: $${data.creditLimit.toFixed(2)}`);
    }
    if (data.minimumPayment !== undefined) {
      summary.push(`Minimum Payment: $${data.minimumPayment.toFixed(2)}`);
    }
    if (data.dueDate) {
      summary.push(`Due Date: ${data.dueDate.toLocaleDateString()}`);
    }

    return summary.length > 0 ? summary.join('\n') : 'No statement data extracted';
  }
}