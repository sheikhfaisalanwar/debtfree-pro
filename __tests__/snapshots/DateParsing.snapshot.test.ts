import { PDFParsingService } from '../../src/services/PDFParsingService';
import { DataStoreService } from '../../src/services/DataStoreService';

describe('Date Parsing Snapshot Tests', () => {
  describe('PDFParsingService Date Extraction', () => {
    it('should parse dates correctly from statement text', () => {
      const sampleStatementText = `
        STATEMENT DATE: 01/15/2024
        DUE DATE: 02/15/2024
        PAYMENT DUE DATE: 02/15/2024
        CLOSING DATE: 01/15/2024
      `;

      const result = PDFParsingService.parseCreditCardStatement(sampleStatementText);
      
      expect({
        statementDate: result.statementDate?.toISOString(),
        dueDate: result.dueDate?.toISOString(),
        statementDateValid: result.statementDate instanceof Date && !isNaN(result.statementDate.getTime()),
        dueDateValid: result.dueDate instanceof Date && !isNaN(result.dueDate.getTime())
      }).toMatchSnapshot();
    });

    it('should handle various date formats', () => {
      const testCases = [
        'STATEMENT DATE: 01/15/2024',
        'STATEMENT DATE: 1/15/24', 
        'STATEMENT DATE: 01-15-2024',
        'DUE DATE: 12/31/2024',
        'DUE DATE: 2/28/24'
      ];

      const results = testCases.map(text => {
        const parsed = PDFParsingService.parseCreditCardStatement(text);
        return {
          input: text,
          statementDate: parsed.statementDate?.toISOString(),
          dueDate: parsed.dueDate?.toISOString(),
          statementDateValid: parsed.statementDate instanceof Date && !isNaN(parsed.statementDate.getTime()),
          dueDateValid: parsed.dueDate instanceof Date && !isNaN(parsed.dueDate.getTime())
        };
      });

      expect(results).toMatchSnapshot();
    });

    it('should handle invalid date formats gracefully', () => {
      const invalidDates = [
        'STATEMENT DATE: 13/45/2024', // Invalid month/day
        'DUE DATE: 02/30/2024',       // Invalid day for February
        'STATEMENT DATE: invalid',     // Non-date text
        'DUE DATE: 00/00/0000'        // Zero dates
      ];

      const results = invalidDates.map(text => {
        const parsed = PDFParsingService.parseCreditCardStatement(text);
        return {
          input: text,
          statementDate: parsed.statementDate,
          dueDate: parsed.dueDate,
          hasValidStatementDate: parsed.statementDate instanceof Date && !isNaN(parsed.statementDate.getTime()),
          hasValidDueDate: parsed.dueDate instanceof Date && !isNaN(parsed.dueDate.getTime())
        };
      });

      expect(results).toMatchSnapshot();
    });
  });

  describe('DataStoreService Date Serialization', () => {
    it('should serialize dates correctly to JSON', () => {
      const testDate = new Date('2024-01-15T10:30:00.000Z');
      const testData = {
        version: '2.0.0',
        accounts: [{
          id: '1',
          name: 'Test Account',
          type: 'CREDIT_CARD',
          institution: 'Test Bank',
          createdDate: testDate,
          lastUpdated: testDate
        }],
        balances: [{
          id: 'bal_1',
          accountId: '1',
          balance: 1000,
          minimumPayment: 50,
          interestRate: 18.99,
          balanceDate: testDate,
          lastUpdated: testDate
        }],
        documents: [{
          id: 'doc_1',
          fileName: 'test.pdf',
          filePath: '/path/to/test.pdf',
          fileType: 'PDF_STATEMENT',
          fileSize: 1024,
          uploadDate: testDate,
          lastModified: testDate,
          processed: false,
          processingStatus: 'PENDING'
        }],
        statements: [{
          id: 'stmt_1',
          accountId: '1',
          statementDate: testDate,
          dueDate: testDate,
          balance: 1000,
          minimumPayment: 50,
          interestCharged: 15,
          transactions: [{
            id: 'txn_1',
            date: testDate,
            amount: 100,
            description: 'Test Transaction',
            type: 'purchase'
          }],
          payments: [{
            id: 'pay_1',
            date: testDate,
            amount: 200,
            description: 'Test Payment',
            type: 'payment'
          }],
          importedDate: testDate,
          lastUpdated: testDate
        }],
        settings: {
          extraPayment: 100,
          strategy: 'SNOWBALL',
          currency: 'CAD',
          dateFormat: 'YYYY-MM-DD',
          notifications: {
            enabled: true,
            dueDateReminders: true,
            paymentReminders: true,
            reminderDays: 3
          }
        },
        metadata: {
          createdDate: testDate,
          lastUpdated: testDate,
          dataVersion: '2.0.0',
          totalMigrations: 0
        }
      };

      // Test serialization
      const serialized = (DataStoreService as any).serializeDataStore(testData);
      
      expect({
        serializedType: typeof serialized,
        accountCreatedDate: serialized.accounts[0].createdDate,
        balanceDate: serialized.balances[0].balanceDate,
        documentUploadDate: serialized.documents[0].uploadDate,
        statementDate: serialized.statements[0].statementDate,
        transactionDate: serialized.statements[0].transactions[0].date,
        paymentDate: serialized.statements[0].payments[0].date,
        metadataCreatedDate: serialized.metadata.createdDate
      }).toMatchSnapshot();
    });

    it('should parse dates correctly from JSON', () => {
      const serializedData = {
        version: '2.0.0',
        accounts: [{
          id: '1',
          name: 'Test Account',
          type: 'CREDIT_CARD',
          institution: 'Test Bank',
          createdDate: '2024-01-15T10:30:00.000Z',
          lastUpdated: '2024-01-15T10:30:00.000Z'
        }],
        balances: [{
          id: 'bal_1',
          accountId: '1',
          balance: 1000,
          minimumPayment: 50,
          interestRate: 18.99,
          balanceDate: '2024-01-15T10:30:00.000Z',
          lastUpdated: '2024-01-15T10:30:00.000Z'
        }],
        documents: [{
          id: 'doc_1',
          fileName: 'test.pdf',
          filePath: '/path/to/test.pdf',
          fileType: 'PDF_STATEMENT',
          fileSize: 1024,
          uploadDate: '2024-01-15T10:30:00.000Z',
          lastModified: '2024-01-15T10:30:00.000Z',
          processed: false,
          processingStatus: 'PENDING'
        }],
        statements: [{
          id: 'stmt_1',
          accountId: '1',
          statementDate: '2024-01-15T10:30:00.000Z',
          dueDate: '2024-02-15T10:30:00.000Z',
          balance: 1000,
          minimumPayment: 50,
          interestCharged: 15,
          transactions: [{
            id: 'txn_1',
            date: '2024-01-15T10:30:00.000Z',
            amount: 100,
            description: 'Test Transaction',
            type: 'purchase'
          }],
          payments: [{
            id: 'pay_1',
            date: '2024-01-15T10:30:00.000Z',
            amount: 200,
            description: 'Test Payment',
            type: 'payment'
          }],
          importedDate: '2024-01-15T10:30:00.000Z',
          lastUpdated: '2024-01-15T10:30:00.000Z'
        }],
        settings: {
          extraPayment: 100,
          strategy: 'SNOWBALL',
          currency: 'CAD',
          dateFormat: 'YYYY-MM-DD',
          notifications: {
            enabled: true,
            dueDateReminders: true,
            paymentReminders: true,
            reminderDays: 3
          }
        },
        metadata: {
          createdDate: '2024-01-15T10:30:00.000Z',
          lastUpdated: '2024-01-15T10:30:00.000Z',
          dataVersion: '2.0.0',
          totalMigrations: 0
        }
      };

      // Test parsing
      const parsed = (DataStoreService as any).parseDataStore(serializedData);
      
      expect({
        accountCreatedDateType: typeof parsed.accounts[0].createdDate,
        accountCreatedDateValid: parsed.accounts[0].createdDate instanceof Date && !isNaN(parsed.accounts[0].createdDate.getTime()),
        balanceDateType: typeof parsed.balances[0].balanceDate,
        balanceDateValid: parsed.balances[0].balanceDate instanceof Date && !isNaN(parsed.balances[0].balanceDate.getTime()),
        documentUploadDateType: typeof parsed.documents[0].uploadDate,
        documentUploadDateValid: parsed.documents[0].uploadDate instanceof Date && !isNaN(parsed.documents[0].uploadDate.getTime()),
        statementDateType: typeof parsed.statements[0].statementDate,
        statementDateValid: parsed.statements[0].statementDate instanceof Date && !isNaN(parsed.statements[0].statementDate.getTime()),
        transactionDateType: typeof parsed.statements[0].transactions[0].date,
        transactionDateValid: parsed.statements[0].transactions[0].date instanceof Date && !isNaN(parsed.statements[0].transactions[0].date.getTime()),
        paymentDateType: typeof parsed.statements[0].payments[0].date,
        paymentDateValid: parsed.statements[0].payments[0].date instanceof Date && !isNaN(parsed.statements[0].payments[0].date.getTime()),
        metadataCreatedDateType: typeof parsed.metadata.createdDate,
        metadataCreatedDateValid: parsed.metadata.createdDate instanceof Date && !isNaN(parsed.metadata.createdDate.getTime())
      }).toMatchSnapshot();
    });

    it('should handle invalid date strings gracefully', () => {
      const invalidDateData = {
        version: '2.0.0',
        accounts: [{
          id: '1',
          name: 'Test Account',
          type: 'CREDIT_CARD',
          institution: 'Test Bank',
          createdDate: 'invalid-date-string',
          lastUpdated: 'another-invalid-date'
        }],
        balances: [{
          id: 'bal_1',
          accountId: '1',
          balance: 1000,
          minimumPayment: 50,
          interestRate: 18.99,
          balanceDate: 'not-a-date',
          lastUpdated: 'also-not-a-date'
        }],
        documents: [],
        statements: [],
        settings: {
          extraPayment: 100,
          strategy: 'SNOWBALL',
          currency: 'CAD',
          dateFormat: 'YYYY-MM-DD',
          notifications: {
            enabled: true,
            dueDateReminders: true,
            paymentReminders: true,
            reminderDays: 3
          }
        },
        metadata: {
          createdDate: 'bad-date',
          lastUpdated: 'worse-date',
          dataVersion: '2.0.0',
          totalMigrations: 0
        }
      };

      try {
        const parsed = (DataStoreService as any).parseDataStore(invalidDateData);
        
        expect({
          accountCreatedDateValid: parsed.accounts[0].createdDate instanceof Date && !isNaN(parsed.accounts[0].createdDate.getTime()),
          accountLastUpdatedValid: parsed.accounts[0].lastUpdated instanceof Date && !isNaN(parsed.accounts[0].lastUpdated.getTime()),
          balanceDateValid: parsed.balances[0].balanceDate instanceof Date && !isNaN(parsed.balances[0].balanceDate.getTime()),
          balanceLastUpdatedValid: parsed.balances[0].lastUpdated instanceof Date && !isNaN(parsed.balances[0].lastUpdated.getTime()),
          metadataCreatedDateValid: parsed.metadata.createdDate instanceof Date && !isNaN(parsed.metadata.createdDate.getTime()),
          metadataLastUpdatedValid: parsed.metadata.lastUpdated instanceof Date && !isNaN(parsed.metadata.lastUpdated.getTime())
        }).toMatchSnapshot();
      } catch (error) {
        expect({
          error: error instanceof Error ? error.message : 'Unknown error',
          errorType: typeof error
        }).toMatchSnapshot();
      }
    });
  });
});