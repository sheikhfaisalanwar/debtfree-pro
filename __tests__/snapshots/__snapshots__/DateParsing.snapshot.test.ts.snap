// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`Date Parsing Snapshot Tests DataStoreService Date Serialization should handle invalid date strings gracefully 1`] = `
{
  "accountCreatedDateValid": false,
  "accountLastUpdatedValid": false,
  "balanceDateValid": false,
  "balanceLastUpdatedValid": false,
  "metadataCreatedDateValid": false,
  "metadataLastUpdatedValid": false,
}
`;

exports[`Date Parsing Snapshot Tests DataStoreService Date Serialization should parse dates correctly from JSON 1`] = `
{
  "accountCreatedDateType": "object",
  "accountCreatedDateValid": true,
  "balanceDateType": "object",
  "balanceDateValid": true,
  "documentUploadDateType": "object",
  "documentUploadDateValid": true,
  "metadataCreatedDateType": "object",
  "metadataCreatedDateValid": true,
  "paymentDateType": "object",
  "paymentDateValid": true,
  "statementDateType": "object",
  "statementDateValid": true,
  "transactionDateType": "object",
  "transactionDateValid": true,
}
`;

exports[`Date Parsing Snapshot Tests DataStoreService Date Serialization should serialize dates correctly to JSON 1`] = `
{
  "accountCreatedDate": "2024-01-15T10:30:00.000Z",
  "balanceDate": "2024-01-15T10:30:00.000Z",
  "documentUploadDate": "2024-01-15T10:30:00.000Z",
  "metadataCreatedDate": "2024-01-15T10:30:00.000Z",
  "paymentDate": "2024-01-15T10:30:00.000Z",
  "serializedType": "object",
  "statementDate": "2024-01-15T10:30:00.000Z",
  "transactionDate": "2024-01-15T10:30:00.000Z",
}
`;

exports[`Date Parsing Snapshot Tests PDFParsingService Date Extraction should handle invalid date formats gracefully 1`] = `
[
  {
    "dueDate": undefined,
    "hasValidDueDate": false,
    "hasValidStatementDate": false,
    "input": "STATEMENT DATE: 13/45/2024",
    "statementDate": undefined,
  },
  {
    "dueDate": 2024-03-01T05:00:00.000Z,
    "hasValidDueDate": true,
    "hasValidStatementDate": false,
    "input": "DUE DATE: 02/30/2024",
    "statementDate": undefined,
  },
  {
    "dueDate": undefined,
    "hasValidDueDate": false,
    "hasValidStatementDate": false,
    "input": "STATEMENT DATE: invalid",
    "statementDate": undefined,
  },
  {
    "dueDate": undefined,
    "hasValidDueDate": false,
    "hasValidStatementDate": false,
    "input": "DUE DATE: 00/00/0000",
    "statementDate": undefined,
  },
]
`;

exports[`Date Parsing Snapshot Tests PDFParsingService Date Extraction should handle various date formats 1`] = `
[
  {
    "dueDate": undefined,
    "dueDateValid": false,
    "input": "STATEMENT DATE: 01/15/2024",
    "statementDate": "2024-01-15T05:00:00.000Z",
    "statementDateValid": true,
  },
  {
    "dueDate": undefined,
    "dueDateValid": false,
    "input": "STATEMENT DATE: 1/15/24",
    "statementDate": "2024-01-15T05:00:00.000Z",
    "statementDateValid": true,
  },
  {
    "dueDate": undefined,
    "dueDateValid": false,
    "input": "STATEMENT DATE: 01-15-2024",
    "statementDate": "2024-01-15T05:00:00.000Z",
    "statementDateValid": true,
  },
  {
    "dueDate": "2024-12-31T05:00:00.000Z",
    "dueDateValid": true,
    "input": "DUE DATE: 12/31/2024",
    "statementDate": undefined,
    "statementDateValid": false,
  },
  {
    "dueDate": "2024-02-28T05:00:00.000Z",
    "dueDateValid": true,
    "input": "DUE DATE: 2/28/24",
    "statementDate": undefined,
    "statementDateValid": false,
  },
]
`;

exports[`Date Parsing Snapshot Tests PDFParsingService Date Extraction should parse dates correctly from statement text 1`] = `
{
  "dueDate": "2024-02-15T05:00:00.000Z",
  "dueDateValid": true,
  "statementDate": "2024-01-15T05:00:00.000Z",
  "statementDateValid": true,
}
`;
